src/MPDeviceBleImpl.cpp

```diff
@@ -480,6 +480,68 @@ void MPDeviceBleImpl::processDebugMsg(const QByteArray &data, bool &isDebugMsg)
     }
 }
 
+QByteArray MPDeviceBleImpl::getFreeAddress(const int virtualAddr)
+{
+    if (m_parentAddrMapping.contains(virtualAddr))
+    {
+        return m_parentAddrMapping[virtualAddr];
+    }
+    else if (m_childAddrMapping.contains(virtualAddr))
+    {
+        return m_childAddrMapping[virtualAddr];
+    }
+    qCritical() << "No address loaded for virtual address: " << virtualAddr;
+    return QByteArray{};
+}
+
+void MPDeviceBleImpl::loadFreeAddresses(AsyncJobs *jobs, const QByteArray &addressFrom, const MPDeviceProgressCb &cbProgress)
+{
+    Q_UNUSED(cbProgress)
+    if (0 == m_parentNodeNeeded && 0 == m_childNodeNeeded)
+    {
+        qDebug() << "No free address is required";
+        return;
+    }
+    auto addressPackage = addressFrom;
+    addressPackage.append(bleProt->toLittleEndianFromInt(m_parentNodeNeeded));
+    addressPackage.append(bleProt->toLittleEndianFromInt(m_childNodeNeeded));
+    jobs->append(new MPCommandJob(mpDev, MPCmd::GET_FREE_ADDRESSES,
+                                  addressPackage,
+                                  [this](const QByteArray &data, bool &) -> bool
+        {
+            const auto msgSize = bleProt->getMessageSize(data);
+            if ((m_parentNodeNeeded + m_childNodeNeeded)*2 != msgSize)
```

Comment:
maybe replace 2 with a define or something :D.
I think you still need to tackle this: https://github.com/mooltipass/minible/blob/master/source_code/main_mcu/src/COMMS/comms_hid_msgs.c#L464 , correct?

Reply:
Switched to ADDRESS_LENGTH instead of 2.
Well I think it is handled with `(m_parentNodeNeeded + m_childNodeNeeded) * MPNode::ADDRESS_LENGTH != msgSize`, because in case of error the payload size will be 1, according to fw code, and then this will be true and the job returns with false.
But I can add an extra 1 check for msgSize too.